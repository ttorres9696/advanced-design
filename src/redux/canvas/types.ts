import { Vector2d } from 'konva/types/types';

import { ShapeCoords } from '../../shared/components/ShapeElement/types';

export const SELECT_SHAPE = 'SELECT_SHAPE';
export const DESELECT_SHAPE = 'DESELECT_SHAPE';
export const INCREASE_ZOOM = 'INCREASE_ZOOM';
export const DECREASE_ZOOM = 'DECREASE_ZOOM';
export const NORMALIZE_ZOOM = 'NORMALIZE_ZOOM';
export const SET_CANVAS_ORIGIN = 'SET_CANVAS_ORIGIN';
export const INCREMENT_TOTAL_SHAPES = 'INCREMENT_TOTAL_SHAPES';
export const DECREMENT_TOTAL_SHAPES = 'DECREMENT_TOTAL_SHAPES';
export const SET_NEW_DRAGGABLE_POINT_MODE = 'SET_NEW_DRAGGABLE_POINT_MODE';
export const SET_MOUSE_POSITION = 'SET_MOUSE_POSITION';
export const SET_DELETE_DRAGGABLE_POINT_MODE = 'SET_DELETE_DRAGGABLE_POINT_MODE';
export const TOGGLE_MAGNET_MODE = 'TOGGLE_MAGNET_MODE';
export const SET_DRAGGING = 'SET_DRAGGING';
export const SET_SELECT_MODE = 'SET_SELECT_MODE';

export const originCoords = 20;
export type SelectionType = 'reset' | 'add';

interface SelectShapeAction {
  type: typeof SELECT_SHAPE;
  payload: {
    solarElementId: string;
    layer: string;
    selectionType: SelectionType;
  };
}

interface DeselectShapeAction {
  type: typeof DESELECT_SHAPE;
  payload: {
    solarElementId: string;
  };
}

interface IncreaseZoomAction {
  type: typeof INCREASE_ZOOM;
}

interface DecreaseZoomAction {
  type: typeof DECREASE_ZOOM;
}

interface NormalizeZoomAction {
  type: typeof NORMALIZE_ZOOM;
}

interface SetCanvasOriginAction {
  type: typeof SET_CANVAS_ORIGIN;
  payload: ShapeCoords;
}

interface IncrementTotalShapesAction {
  type: typeof INCREMENT_TOTAL_SHAPES;
}

interface DecrementTotalShapesAction {
  type: typeof DECREMENT_TOTAL_SHAPES;
  payload: number;
}

interface SetNewDraggablePointModeAction {
  type: typeof SET_NEW_DRAGGABLE_POINT_MODE;
  payload: boolean;
}

interface SetDeleteDraggablePointMode {
  type: typeof SET_DELETE_DRAGGABLE_POINT_MODE;
  payload: boolean;
}

interface SetMousePositionAction {
  type: typeof SET_MOUSE_POSITION;
  payload?: Vector2d;
}

interface ToggleMagnetModeAction {
  type: typeof TOGGLE_MAGNET_MODE;
}

interface SetDragging {
  type: typeof SET_DRAGGING;
  payload: boolean;
}

interface SetSelectMode {
  type: typeof SET_SELECT_MODE;
  payload: boolean;
}

export type CanvasActionTypes =
  | SelectShapeAction
  | DeselectShapeAction
  | IncreaseZoomAction
  | DecreaseZoomAction
  | NormalizeZoomAction
  | SetCanvasOriginAction
  | IncrementTotalShapesAction
  | DecrementTotalShapesAction
  | SetNewDraggablePointModeAction
  | SetDeleteDraggablePointMode
  | SetMousePositionAction
  | ToggleMagnetModeAction
  | SetDragging
  | SetSelectMode;
