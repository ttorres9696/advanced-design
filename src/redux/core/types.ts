import { ModuleSpec } from '../../shared/models/SolarDesign.interface';
import { DeleteDialogParams, WindowDimension } from './state';

export const FETCHED_MODULE_SPECS = 'FETCHED_MODULE_SPECS';
export const SET_DELETE_DIALOG_PARAMS = 'SET_DELETE_DIALOG_PARAMS';
export const SET_WINDOW_DIMENSION = 'SET_WINDOW_DIMENSION';
export const SET_PRE_LOADING = 'SET_PRE_LOADING';
export const CANCEL_PRE_LOADING = 'CANCEL_PRE_LOADING';
export const DECREASE_PRODUCTION_VALUE = 'DECREASE_PRODUCTION_VALUE';
export const INCREASE_PRODUCTION_VALUE = 'INCREASE_PRODUCTION_VALUE';
export const RESET_PRODUCTION_VALUE = 'RESET_PRODUCTION_VALUE';

interface FetchedModuleSpecsAction {
  type: typeof FETCHED_MODULE_SPECS;
  payload: ModuleSpec[];
}

interface SetDeleteDialogParamsAction {
  type: typeof SET_DELETE_DIALOG_PARAMS;
  payload: DeleteDialogParams;
}

interface SetWindowDimensionAction {
  type: typeof SET_WINDOW_DIMENSION;
  payload: WindowDimension;
}

interface SetPreLoadingAction {
  type: typeof SET_PRE_LOADING;
  payload: boolean;
}

interface CancelPreLoading {
  type: typeof CANCEL_PRE_LOADING;
}

interface DecreaseProductionValue {
  type: typeof DECREASE_PRODUCTION_VALUE;
  payload: number;
}

interface IncreaseProductionValue {
  type: typeof INCREASE_PRODUCTION_VALUE;
  payload: number;
}

interface ResetProductionValue {
  type: typeof RESET_PRODUCTION_VALUE;
}

export type CoreActionTypes =
  | FetchedModuleSpecsAction
  | SetDeleteDialogParamsAction
  | SetWindowDimensionAction
  | SetPreLoadingAction
  | CancelPreLoading
  | DecreaseProductionValue
  | IncreaseProductionValue
  | ResetProductionValue;
