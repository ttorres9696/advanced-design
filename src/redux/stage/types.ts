import { ShapeCoords } from '../../shared/components/ShapeElement/types';
import { ModuleSpec, SolarDesign } from '../../shared/models/SolarDesign.interface';
import { SolarElement } from '../../shared/models/SolarElement.interface';
import { SelectionType } from '../canvas/types';

export const ADD_SHAPE = 'ADD_SHAPE';
export const UPDATE_SHAPE = 'UPDATE_SHAPE';
export const REMOVE_SHAPE = 'REMOVE_SHAPE';
export const TOGGLE_VISIBILITY = 'TOGGLE_VISIBILITY';
export const TOGGLE_LOCK = 'TOGGLE_LOCK';
export const IMPORT_DATA = 'IMPORT_DATA';
export const UNDO_HISTORY = 'UNDO_HISTORY';
export const REDO_HISTORY = 'REDO_HISTORY';
export const RESET_HISTORY = 'RESET_HISTORY';
export const START_HISTORY_ITEM = 'START_HISTORY_ITEM';
export const FINISH_HISTORY_ITEM = 'FINISH_HISTORY_ITEM';
export const ADD_NEW_DRAGGABLE_POINT = 'ADD_NEW_DRAGGABLE_POINT';
export const REMOVE_DRAGGABLE_POINT = 'REMOVE_DRAGGABLE_POINT';
export const SET_SELECTED_ELEMENT = 'SET_SELECTED_ELEMENT';
export const SET_SELECTED_MODULE_SPEC = 'SET_SELECTED_MODULE_SPEC';
export const SET_TRANSFORMER_NODES = 'SET_TRANSFORMER_NODES';
export const CLEAR_TRANSFORMER_NODES_STATE = 'CLEAR_TRANSFORMER_NODES_STATE';
export const TOGGLE_MODULE = 'TOGGLE_MODULE';

export const defaultMultiplierScale = 10;

export interface SolarElementReference {
  layer?: string;
  id: string;
}

interface AddShapeAction {
  type: typeof ADD_SHAPE;
  payload: {
    element: SolarElement;
    moduleSeries: string;
  };
}

interface UpdateShapeAction {
  type: typeof UPDATE_SHAPE;
  payload: SolarElement;
}

interface RemoveShapeAction {
  type: typeof REMOVE_SHAPE;
  payload: {
    layer: string;
    elementIds: string[];
  };
}

interface ToggleVisibilityAction {
  type: typeof TOGGLE_VISIBILITY;
  payload: SolarElementReference;
}

interface ToggleLockAction {
  type: typeof TOGGLE_LOCK;
  payload: SolarElementReference;
}

interface ImportDataAction {
  type: typeof IMPORT_DATA;
  payload: SolarDesign;
}

interface UndoHistoryAction {
  type: typeof UNDO_HISTORY;
}

interface RedoHistoryAction {
  type: typeof REDO_HISTORY;
}

interface ResetHistoryAction {
  type: typeof RESET_HISTORY;
}

interface StartHistoryItemAction {
  type: typeof START_HISTORY_ITEM;
  payload: {
    ids: string[];
    layer: string;
  };
}

interface FinishHistoryItemAction {
  type: typeof FINISH_HISTORY_ITEM;
  payload: {
    ids: string[];
    layer: string;
  };
}

interface AddNewDraggablePoint {
  type: typeof ADD_NEW_DRAGGABLE_POINT;
  payload: {
    layer: string;
    elementId: string;
    newPoint: ShapeCoords;
  };
}

interface RemoveDraggablePoint {
  type: typeof REMOVE_DRAGGABLE_POINT;
  payload: {
    layer: string;
    elementId: string;
    pointIndex: number;
  };
}

interface SetSelectedElement {
  type: typeof SET_SELECTED_ELEMENT;
  payload: {
    layer: string;
    elementId: string;
    selected: boolean;
  };
}

interface SetSelectedModuleSpec {
  type: typeof SET_SELECTED_MODULE_SPEC;
  payload: ModuleSpec;
}

interface SetTransformerNodes {
  type: typeof SET_TRANSFORMER_NODES;
  payload: {
    selectionType: SelectionType;
    elementIds: string[];
  };
}

interface ClearTransformerNodesState {
  type: typeof CLEAR_TRANSFORMER_NODES_STATE;
}

interface ToggleModule {
  type: typeof TOGGLE_MODULE;
  payload: {
    elementId: string;
  };
}

export type StageActionTypes =
  | AddShapeAction
  | UpdateShapeAction
  | RemoveShapeAction
  | ToggleVisibilityAction
  | ToggleLockAction
  | ImportDataAction
  | UndoHistoryAction
  | RedoHistoryAction
  | ResetHistoryAction
  | StartHistoryItemAction
  | FinishHistoryItemAction
  | AddNewDraggablePoint
  | RemoveDraggablePoint
  | SetSelectedElement
  | SetSelectedModuleSpec
  | SetTransformerNodes
  | ClearTransformerNodesState
  | ToggleModule;
