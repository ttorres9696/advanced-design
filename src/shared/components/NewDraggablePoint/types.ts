import { Vector2d } from 'konva/types/types';

export interface NewDraggablePointProps {
  newDraggablePointMode: boolean;
  mousePosition?: Vector2d;
}
