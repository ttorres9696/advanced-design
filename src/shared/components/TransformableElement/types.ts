import Konva from 'konva';
import { Node } from 'konva/types/Node';
import { ReactElement } from 'react';

import { SelectionType } from '../../../redux/canvas/types';
import { ShapeElementType } from '../ShapeElement/types';

export interface UseTansfromableElementProps {
  id: string;
  layer: string;
  onStartChange: (id: string, layer: string) => void;
  onChange: (id: string, layer: string, params: TransformableElementChangeParams) => void;
  onFinishChange: (id: string, layer: string) => void;
  onRotate: (id: string, layer: string, pointsRotationInRadians: number[], points: number[][]) => void;
  setDragging: (dragging: boolean) => void;
  version: number;
  deleteDraggablePointMode: boolean;
  onMoveStart: (layer: string, node: Node) => void;
  onMove: (id: string, layer: string, node: Node, e: Konva.KonvaEventObject<DragEvent>) => void;
  onMoveEnd: (layer: string, node: Node) => void;
  onSelect: (id: string, layer: string, selectionType: SelectionType, node: Node) => void;
  visible: boolean;
  locked: boolean;
  multiple: boolean;
  selectMode: boolean;
}

export interface TransformableElementProps extends UseTansfromableElementProps {
  shapeType: ShapeElementType;
  resizeEnabled?: boolean;
  children?: ReactElement | ReactElement[];
  isSelected: boolean;
  active: boolean;
}

export interface TransformableElementChangeParams {
  x?: number;
  y?: number;
  scaleX?: number;
  scaleY?: number;
  width?: number;
  height?: number;
  radius?: number;
  points?: number[][];
  pointsRotationInRadians?: number[];
}

export interface TransformerDimensions {
  width: number;
  height: number;
}
