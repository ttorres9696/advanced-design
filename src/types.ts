import { CallbacksProps } from './shared/models/CallbacksProps.interface';
import { SolarDesign } from './shared/models/SolarDesign.interface';

export interface AdvancedDesignProps extends CallbacksProps {
  solarDesign: SolarDesign;
}
