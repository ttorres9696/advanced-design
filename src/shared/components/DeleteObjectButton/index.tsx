import { Button } from '@material-ui/core';
import { Delete as DeleteIcon } from '@material-ui/icons';
import React from 'react';

import useStyles from './styles';
import { DeleteObjectButtonProps } from './types';

const DeleteObjectButton: React.FC<DeleteObjectButtonProps> = ({ onDelete, disabled, multiple }) => {
  const classes = useStyles();

  return (
    <Button
      variant="outlined"
      startIcon={<DeleteIcon />}
      fullWidth
      className={classes.deleteButton}
      onClick={onDelete}
      size="small"
      disabled={disabled}
    >
      Delete Object{(multiple && 's') || ''}
    </Button>
  );
};

export default React.memo(DeleteObjectButton);
