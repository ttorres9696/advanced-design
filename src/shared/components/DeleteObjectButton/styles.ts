import RED from '@material-ui/core/colors/red';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';

export default makeStyles((theme: Theme) =>
  createStyles({
    deleteButton: {
      borderColor: RED[200],
      color: RED[500],
      '&:hover': {
        borderColor: RED[500],
        backgroundColor: RED[50],
      },
    },
  }),
);
