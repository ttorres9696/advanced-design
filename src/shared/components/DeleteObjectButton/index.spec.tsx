import { mount, ReactWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import DeleteObjectButton from '.';
import { DeleteObjectButtonProps } from './types';

let wrapper: ReactWrapper;

beforeEach(() => {
  const props: DeleteObjectButtonProps = {
    onDelete: () => {},
    disabled: false,
  };

  wrapper = mount(<DeleteObjectButton {...props} />, {});
});

describe('DeleteObjectButton', () => {
  it('renders correctly', () => expect(toJson(wrapper)).toMatchSnapshot());
});
