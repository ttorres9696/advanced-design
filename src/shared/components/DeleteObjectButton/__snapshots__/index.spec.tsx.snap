// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteObjectButton renders correctly 1`] = `
<Memo(DeleteObjectButton)
  disabled={false}
  onDelete={[Function]}
>
  <WithStyles(ForwardRef(Button))
    className="makeStyles-deleteButton-1"
    disabled={false}
    fullWidth={true}
    onClick={[Function]}
    size="small"
    startIcon={<UNDEFINED />}
    variant="outlined"
  >
    <ForwardRef(Button)
      className="makeStyles-deleteButton-1"
      classes={
        Object {
          "colorInherit": "MuiButton-colorInherit",
          "contained": "MuiButton-contained",
          "containedPrimary": "MuiButton-containedPrimary",
          "containedSecondary": "MuiButton-containedSecondary",
          "containedSizeLarge": "MuiButton-containedSizeLarge",
          "containedSizeSmall": "MuiButton-containedSizeSmall",
          "disableElevation": "MuiButton-disableElevation",
          "disabled": "Mui-disabled",
          "endIcon": "MuiButton-endIcon",
          "focusVisible": "Mui-focusVisible",
          "fullWidth": "MuiButton-fullWidth",
          "iconSizeLarge": "MuiButton-iconSizeLarge",
          "iconSizeMedium": "MuiButton-iconSizeMedium",
          "iconSizeSmall": "MuiButton-iconSizeSmall",
          "label": "MuiButton-label",
          "outlined": "MuiButton-outlined",
          "outlinedPrimary": "MuiButton-outlinedPrimary",
          "outlinedSecondary": "MuiButton-outlinedSecondary",
          "outlinedSizeLarge": "MuiButton-outlinedSizeLarge",
          "outlinedSizeSmall": "MuiButton-outlinedSizeSmall",
          "root": "MuiButton-root",
          "sizeLarge": "MuiButton-sizeLarge",
          "sizeSmall": "MuiButton-sizeSmall",
          "startIcon": "MuiButton-startIcon",
          "text": "MuiButton-text",
          "textPrimary": "MuiButton-textPrimary",
          "textSecondary": "MuiButton-textSecondary",
          "textSizeLarge": "MuiButton-textSizeLarge",
          "textSizeSmall": "MuiButton-textSizeSmall",
        }
      }
      disabled={false}
      fullWidth={true}
      onClick={[Function]}
      size="small"
      startIcon={<UNDEFINED />}
      variant="outlined"
    >
      <WithStyles(ForwardRef(ButtonBase))
        className="MuiButton-root MuiButton-outlined makeStyles-deleteButton-1 MuiButton-outlinedSizeSmall MuiButton-sizeSmall MuiButton-fullWidth"
        component="button"
        disabled={false}
        focusRipple={true}
        focusVisibleClassName="Mui-focusVisible"
        onClick={[Function]}
        type="button"
      >
        <ForwardRef(ButtonBase)
          className="MuiButton-root MuiButton-outlined makeStyles-deleteButton-1 MuiButton-outlinedSizeSmall MuiButton-sizeSmall MuiButton-fullWidth"
          classes={
            Object {
              "disabled": "Mui-disabled",
              "focusVisible": "Mui-focusVisible",
              "root": "MuiButtonBase-root",
            }
          }
          component="button"
          disabled={false}
          focusRipple={true}
          focusVisibleClassName="Mui-focusVisible"
          onClick={[Function]}
          type="button"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-deleteButton-1 MuiButton-outlinedSizeSmall MuiButton-sizeSmall MuiButton-fullWidth"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              <span
                className="MuiButton-startIcon MuiButton-iconSizeSmall"
              >
                <ForwardRef>
                  <WithStyles(ForwardRef(SvgIcon))>
                    <ForwardRef(SvgIcon)
                      classes={
                        Object {
                          "colorAction": "MuiSvgIcon-colorAction",
                          "colorDisabled": "MuiSvgIcon-colorDisabled",
                          "colorError": "MuiSvgIcon-colorError",
                          "colorPrimary": "MuiSvgIcon-colorPrimary",
                          "colorSecondary": "MuiSvgIcon-colorSecondary",
                          "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                          "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                          "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                          "root": "MuiSvgIcon-root",
                        }
                      }
                    >
                      <svg
                        aria-hidden="true"
                        className="MuiSvgIcon-root"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                        />
                      </svg>
                    </ForwardRef(SvgIcon)>
                  </WithStyles(ForwardRef(SvgIcon))>
                </ForwardRef>
              </span>
              Delete Object
            </span>
            <NoSsr>
              <WithStyles(undefined)
                center={false}
              >
                <ForwardRef(TouchRipple)
                  center={false}
                  classes={
                    Object {
                      "child": "MuiTouchRipple-child",
                      "childLeaving": "MuiTouchRipple-childLeaving",
                      "childPulsate": "MuiTouchRipple-childPulsate",
                      "ripple": "MuiTouchRipple-ripple",
                      "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                      "rippleVisible": "MuiTouchRipple-rippleVisible",
                      "root": "MuiTouchRipple-root",
                    }
                  }
                >
                  <span
                    className="MuiTouchRipple-root"
                  >
                    <TransitionGroup
                      childFactory={[Function]}
                      component={null}
                      exit={true}
                    />
                  </span>
                </ForwardRef(TouchRipple)>
              </WithStyles(undefined)>
            </NoSsr>
          </button>
        </ForwardRef(ButtonBase)>
      </WithStyles(ForwardRef(ButtonBase))>
    </ForwardRef(Button)>
  </WithStyles(ForwardRef(Button))>
</Memo(DeleteObjectButton)>
`;
